### ISSUE TITLE
Examle using CNTK for Object Detection ?

### ISSUE THREAD

**Issue by JohnCraigPublic** [Microsoft/CNTK#760]
I'd like to see an example of how one sets up the config files to use CNTK on an "object detection" problem in which you want to classify the object but also compute a bounding box around it.  I imagine this is done with supervised learning - with the system given tons of example images with correct bounding boxes on them.  So, I'd like to see how this labeled data is read in, and what type of network produces the classification and the bounding box output.   I love simple examples that show the essence of things, so in fact, to make it simpler, it could be a CNN that doesn't even do classification -- its fed images known to have a certain object in them, the goal is just to find the bounding box around that object.  What is the simplest way?  Thanks!

**Comment by pkranen** [comment]
Hi John, 

To predict bounding boxes on images you can train a network that takes an image as input and predicts four floating point values. Usually these values are (x, y, w, h) which are specified relative to the image size. To do this in CNTK you need to 

	a) Define both the image and the ground truth regression labels as inputs to your network
	b) Define a network that predicts a bounding box as four values
	c) Define a loss function that compares the predicted bbox with the ground truth
	d) Adapt the reader section in your .cntk config file to read both image and regression labels

As I don't have a bbox regression data set at hand, I provide a short example using the CIFAR-10 data set. The regression labels in the example are the average rgb values of an image normalized to [0, 1]. The script to generate the data (updated CifarConverter.py) as well as the full ndl and .cntk files are in the branch 'imageRegressionExamples'. I'll just highlight the parts corresponding to a-d above.

For a):

```
    features = ImageInput(ImageW, ImageH, ImageC, tag = feature, imageLayout = $imageLayout$)
    regrLabels = Input(LabelDim, tag = label)

```
For b) and c) a very simple network would be the following:

```
DNN=[
    hiddenDim = 64
    h1 = DnnImageLastLayer(ImageW, ImageH, ImageC, hiddenDim, features, 1, 0)
    ol = DNNLastLayer(hiddenDim, LabelDim, h1, 1, 0)
    
    # define regression loss
    sqerr = SquareError(regrLabels, ol)
    
    criterionNodes  = (sqerr)
    evaluationNodes = (sqerr)
    
    OutputNodes = ol
]

```
For d) the reader section in the example looks like this:

```
    reader = [
        randomize = false
        verbosity = 2
        deserializers = (
        [
            type = "CNTKTextFormatDeserializer"
            module = "CNTKTextFormatReader"
            file = "$DataDir$/cifar-10-batches-py/train_regrLabels.txt"
            input = [
                regrLabels = [
                    dim = 3
                    format = "dense"
                ]
            ]
        ]:[
            type = "ImageDeserializer"
            module = "ImageReader"
            file = "$DataDir$/cifar-10-batches-py/train_map.txt"
            input = [
                features = [
                    transforms = (
                    [
                        type = "Scale"
                        width = 32
                        height = 32
                        channels = 3
                        interpolations = "linear"
                    ]:[
                        type = "Mean"
                        meanFile = "$DataDir$/cifar-10-batches-py/CIFAR-10_mean.xml"
                    ]
                    )
                ]
                ignored=[
                    labelDim=10
                ]
            ]
        ]
        )
    ]
```


Note that the network defined above is of course too simple for bounding box regression. In the literature often the output of the last convolutional layer, the so called conv feature map, is used as input to one or more fully connected layers to do bounding box regression (see for example Fast R-CNN at [http://arxiv.org/abs/1504.08083](http://arxiv.org/abs/1504.08083)). The branch contains a second network that connects a regression head to a the last of three convolutional layers (see 07_RegressionConv.ndl).

**Comment by pkranen** [comment]
I added a wiki page with a proper brain script example here: [https://github.com/Microsoft/CNTK/wiki/Train-a-regression-model-on-images](https://github.com/Microsoft/CNTK/wiki/Train-a-regression-model-on-images)

**Comment by JohnCraigPublic** [comment]
Philipp - thank you very much!!

**Comment by pkranen** [comment]
We added a tutorial on [object detection using Fast R-CNN](https://github.com/Microsoft/CNTK/wiki/Object-Detection-using-Fast-R-CNN).

**Comment by drcrook1** [comment]
Any examples for how to define the readers and label files in python?

**Comment by pkranen** [comment]
There is a branch that contains a preliminary version of Fast R-CNN using the CNTK Python API (pkranen/frcnPythonApi). A2_RunCntk_py3.py in that branch has also the reader definition (note the change in cntk/io/__init__.py though). I'm planning to revive it and merge it to master in the coming weeks.
